ducktape/tests/runner.py

```diff
@@ -14,212 +14,175 @@
-                self.log(logging.INFO, "FAIL: " + err_trace)
-
-                test_status = FAIL
-                summary += err_trace
-
-                self.stop_testing = self.session_context.exit_first or isinstance(e, KeyboardInterrupt)
-
-            finally:
-                self.teardown_single_test(teardown_services=not self.session_context.no_teardown)
+                message = self.socket.recv()
+                event = self.serde.deserialize(message)
+                self.handle(event)
+            except Exception as e:
+                s = str(type(e))
+                s += " " + str(e)
+                print "Exception receiving message:", str(e)
+                continue
 
-                stop_time = time.time()
+        for proc in self.proc_list:
+            proc.join()
 
-                result = TestResult(self.current_test_context, test_status, summary, data, start_time, stop_time)
-                self.results.append(result)
+        return self.results
 
-                self.log(logging.INFO, "Summary: %s" % str(result.summary))
-                self.log(logging.INFO, "Data: %s" % str(result.data))
-                if test_num < len(self.tests):
-                    terminal_width, y = get_terminal_size()
-                    print "~" * int(2 * terminal_width / 3)
+    def run_single_test(self):
+        proc = multiprocessing.Process(
+            target=run_client,
+            args=[
+                self.port,
+                self.current_test_context.logger_name,
+                self.current_test_context.results_dir,
+                self.session_context.debug,
+                self.session_context.max_parallel
+            ])
+        self.proc_list.append(proc)
+        proc.start()
+
+    def handle(self, event):
+        self.request_logger.debug(event)
+
+        if event["event_type"] == Request.READY:
+            self.handle_ready(event)
+        elif event["event_type"] in [Request.RUNNING, Request.SETTING_UP, Request.TEARING_DOWN]:
+            self.handle_lifecycle(event)
+        elif event["event_type"] == Request.FINISHED:
+            self.handle_finished(event)
+        elif event["event_type"] == Request.LOG:
+            self.handle_log(event)
+        else:
+            self.handle_other(event)
 
-                test_reporter = SingleResultFileReporter(result)
-                test_reporter.report()
+    def handle_ready(self, event):
+        self.socket.send(
+            self.serde.serialize(ready_reply(self.session_context, self.current_test_context, self.cluster)))
 
-                self.current_test_context, self.current_test = None, None
+        # Test is "active" once we receive the ready request
+        self.active_tests[event["source_id"]] = event["event_type"]
 
-            if self.stop_testing:
-                break
+    def handle_log(self, event):
+        self.log(event["log_level"], event["message"])
+        self.socket.send(self.serde.serialize(event))
```

Comment:
Why are we sending the event back in reply? Looks like we do this repeatedly. Why wouldn't this just be empty or some simple ack?